File: packing_interface/src/algorithm_templates/nfdh.py
import packing_lib
import json
from typing import List, Tuple

class Packing:
    def solve(self, bin_width: int, rectangles: List[Tuple[int, int, int]]) -> List[Tuple[float, float, int, int]]:
        items = packing_lib.expand_items(rectangles)
        items = packing_lib.sort_by_height(items)
        placements = []
        current_level_y = 0 
        current_level_height = 0
        current_level_used_width = 0

        for item in items:
            width = item["width"]
            height = item["height"]

            if current_level_used_width == 0:
                current_level_height = height

            if current_level_used_width + width <= bin_width:
                placements.append((current_level_used_width, current_level_y, width, height))
                current_level_used_width += width
                continue

            current_level_y += current_level_height
            current_level_height = height
            current_level_used_width = width
            placements.append((0, current_level_y, width, height))
        return packing_lib.output_from_placements(bin_width, placements)

File: packing_interface/src/algorithm_templates/nfdh.cpp
#include "packing_lib.h"
using namespace packing;

class Packing {
public:
    std::vector<std::tuple<double, double, int, int>> solve(
        int binWidth,
        const std::vector<std::tuple<int, int, int>>& rectangles
    ) {
        auto items = sort_by_height(expand_items(rectangles));

        std::vector<std::tuple<double, double, int, int>> placements;
        int currentLevelY = 0;
        int currentLevelHeight = 0;
        int currentLevelUsedWidth = 0;

        for (const auto& item : items) {
            if (currentLevelUsedWidth == 0) {
                currentLevelHeight = item.height;
            }

            if (currentLevelUsedWidth + item.width <= binWidth) {
                placements.push_back({
                    static_cast<double>(currentLevelUsedWidth),
                    static_cast<double>(currentLevelY),
                    item.width,
                    item.height,
                });
                currentLevelUsedWidth += item.width;
                continue;
            }

            currentLevelY += currentLevelHeight;
            currentLevelHeight = item.height;
            currentLevelUsedWidth = item.width;
            placements.push_back({0.0, static_cast<double>(currentLevelY), item.width, item.height});
        }

        return placements;
    }
};
